File: 2022040006_김주완/Week03_Recursion/Target_Number.cpp
#include <iostream>
#include <vector>
using namespace std;

int answer = 0;

void dfs(const vector<int>& numbers, int target, int index, int current_sum) {

    if (index == numbers.size()) {
        if (current_sum == target) {
            answer++;
        }
        return;
    }
    dfs(numbers, target, index + 1, current_sum + numbers[index]);
    dfs(numbers, target, index + 1, current_sum - numbers[index]);
}

int solution(vector<int> numbers, int target) {
    answer = 0;
    dfs(numbers, target, 0, 0);
    return answer;
}

int main() {
    vector<int> numbers1 = { 1, 1, 1, 1, 1 };
    int target1 = 3;
    cout << solution(numbers1, target1) << endl;

    vector<int> numbers2 = { 4, 1, 2, 1 };
    int target2 = 4;
    cout << solution(numbers2, target2) << endl;

    return 0;
}

File: 2025049016_문정현/Week03_Recursion/Target_number.py
def solution(numbers, target):
    answer = 0
    
    def dfs(i, value):
        nonlocal answer
        
        if i == len(numbers):
            if value == 0:
                answer += 1
            return

        dfs(i+1, value + numbers[i])
        dfs(i+1, value - numbers[i])
        
    dfs(0, target)
    
    return answer

num = [1, 1, 1, 1, 1]
target = 3
print(solution(num, target))